### DISCUSSION TITLE
How to secure Longhorn UI / API

### DISCUSSION THREAD

**Discussion by remram44** [longhorn/longhorn#3031] [Q&A]
I would like to prevent access to the Longhorn UI to anyone but admins. The UI allows dangerous operations like deleting volumes.

The only documentation I found is [how to add an Ingress with a password](https://longhorn.io/docs/1.2.0/deploy/accessing-the-ui/longhorn-ingress/). This allows for public access to the UI, with a password. However this doesn't prevent direct access to the Service or Pod directly (via clusterIP) which is possible whether or not an Ingress exists.

I think it is a problem if any pod in the cluster can connect to the UI and delete volumes. Is there a way to secure this, short of disabling the UI or locking it with NetworkPolicy?

**Comment by jenting** [comment]
I think you could consider using the [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) to allow the admin have the permission to access the `longhorn-system` namespace.

**Reply by remram44** [reply]
Communication with the UI does not happen through the Kubernetes API server, so that's unrelated.

**Reply by remram44** [reply]
I don't know if I'm not being clear or something. Currently anyone who has permission to run a pod in any namespace can use that pod to talk to the Longhorn UI at http://longhorn-frontend.longhorn-system.svc/. Through that, they can create/delete volumes with no authentication. The role you assigned that user is completely ignored because there is no security on the UI.

**Reply by jenting** [reply]
My first thought is it should be achieved by RBAC, ref to https://stackoverflow.com/a/59063250

If the above did not meet your requirement and you're like to restrict on the network level, probably you could check the `Network Policy` setting on your CNI plugin (it depends on which CNI plugin you're using). To restrict the Pod from other namespaces is not able to access the `longhorn-system` namespace.

**Reply by remram44** [reply]
You are still misunderstanding the problem. This has nothing to do with the Kubernetes API (which has RBAC). The Longhorn UI does not have its own RBAC.

**Reply by jenting** [reply]
What if consider using the Ingress + authentication (e.g. LDAP)?

**Reply by remram44** [reply]
The service is still accessible no matter how many ingresses there are

**Reply by jenting** [reply]
Is this one https://github.com/longhorn/longhorn/issues/1805#issuecomment-812901266 you want to achieve?

**Reply by remram44** [reply]
Yes! I'm trying to achieve any form of security, where I can actually deploy Longhorn to a non-toy cluster.

I hadn't found #1805 (and #1983). I guess I have to uninstall Longhorn and install a different system, since security has not been considered yet.

**Reply by lahwaacz** [reply]
Three years later Longhorn unfortunately still did not improve its security :frowning_face:

**Comment by remram44** [comment]
This should be issued a CVE or something. Anybody can delete all data on any cluster they can run a pod on, or if any container on that cluster or that cluster's network gets a RCE vulnerability, or even just a SSRF vulnerability, your whole storage cluster gets owned.

**Comment by steffeneichler** [comment]
Hi,

our sec pen testers also found this problem. One fix could be to set more restrictive networkpolicies than the rolled out per default.
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: np-default-allow-all
  namespace: longhorn-system
spec:
  egress:
  - {}
  ingress:
  - {}
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
```
